nucliadb/nucliadb/ingest/tests/orm/test_nodes_manager.py

```diff
@@ -0,0 +1,169 @@
+from nucliadb.ingest.orm.nodes_manager import NodesManager
+from nucliadb_utils.keys import KB_SHARDS
+
+
+@pytest.fixture(scope="function")
+async def fake_kbid() -> AsyncIterable[str]:
+    yield uuid.uuid4().hex
+
+
+@pytest.fixture(scope="function")
+async def fake_nodes():
+    await Node.set(
+        "node-0",
+        address="nohost:9999",
+        type=NodeType.IO,
+        load_score=0.0,
+        shard_count=0,
+        dummy=True,
+    )
+    await Node.set(
+        "node-1",
+        address="nohost:9999",
+        type=NodeType.IO,
+        load_score=0.0,
+        shard_count=0,
+        dummy=True,
+    )
+    await Node.set(
+        "node-2",
+        address="nohost:9999",
+        type=NodeType.IO,
+        load_score=0.0,
+        shard_count=0,
+        dummy=True,
+    )
+    yield
+
+
+@pytest.fixture(scope="function")
+async def shards(fake_nodes, fake_kbid: str, redis_driver: Driver):
+    """
+    Shards:
+    - shard-a
+    - shard-b
+
+    ShardReplicas:
+    - shard-a.1 -> node-0
+    - shard-a.2 -> node-1
+    - shard-b.1 -> node-1
+    - shard-b.2 -> node-2
+    """
+    driver = redis_driver
+    kbid = fake_kbid
+
+    shards = Shards()
+    shards.kbid = kbid
+
+    shard_a = ShardObject()
+    shard_a.shard = "shard-a"
+    shard_a.replicas.extend(
+        [
+            ShardReplica(
+                node="node-0",
+                shard=ShardCreated(id="shard-a.1"),
+            ),
+            ShardReplica(
+                node="node-1",
+                shard=ShardCreated(id="shard-a.2"),
+            ),
+        ]
+    )
+
+    shard_b = ShardObject()
+    shard_b.shard = "shard-b"
+    shard_b.replicas.extend(
+        [
+            ShardReplica(
+                node="node-1",
+                shard=ShardCreated(id="shard-b.1"),
+            ),
+            ShardReplica(
+                node="node-2",
+                shard=ShardCreated(id="shard-b.2"),
+            ),
+        ]
+    )
+
+    shards.shards.append(shard_a)
+    shards.shards.append(shard_b)
+
+    key = KB_SHARDS.format(kbid=kbid)
+
+    async with driver.transaction() as txn:
+        await txn.set(key, shards.SerializeToString())
+        await txn.commit(resource=False)
+
+    yield shards
+
+    async with driver.transaction() as txn:
+        await txn.delete(key)
+        await txn.commit(resource=False)
+
+
+@pytest.mark.parametrize(
+    "shard_index,nodes",
+    [
+        (0, {"node-0", "node-1"}),
+        (1, {"node-1", "node-2"}),
+    ],
+)
+@pytest.mark.asyncio
+async def test_choose_node(shards, redis_driver: Driver, shard_index: int, nodes: set):
+    driver = redis_driver
+
+    nodes_manager = NodesManager(driver=driver, cache=mock.Mock())
+    shard = shards.shards[shard_index]
+    node_ids = set()
+    for i in range(100):
+        _, _, node_id = nodes_manager.choose_node(shard)
+        node_ids.add(node_id)
+    assert node_ids == nodes, "Random numbers have defeat this test"
```

Comment:
To simplify the test, Instead of this you could assert that the returned node_id is from one of the replicas of the shard?

Reply:
And how will you test the randomness of the algorithm?